### PR TITLE
Lock async context manager

### PR DESCRIPTION
**Pull request by rhdxmr**
Hello, @cybergrind 

I changed aiozk.recipes.Lock as async context manager.

So Lock recipe can be used as below.

async with zk.recipes.Lock(path):
      ...

Of course, .acquire() coroutine still works as it did.

### CODE REVIEW

**Comment by rhdxmr** [comment]
@cybergrind 
I don't get why this PR includes previous PR (https://github.com/micro-fan/aiozk/pull/65)
It seems that the PR#65 was already merged into master.

**Comment by cybergrind** [comment]
Hey @rhdxmr. PR itself looks good but my eye has caught not very beautiful `BaseLock.make_contextmanager` with dynamic context manager creation inside it and it is never used as a context manager (from what I can see).
Also, that leads to non-beautiful `_acquire2` and `_release2`.

Probably it should resemble the usual python approach when `__aenter__` just calls `acquire` and `__aexit__` calls `release`, like here https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L12.

Thus we could remove `make_contextmanager` completely and split its functionality between `acquire` and `release`
What do you think about that?

**Comment by rhdxmr** [comment]
@cybergrind 
I agree with you.

The behavior of the current `.acquire()` method is weird.  Because it returns async context manager. Besides `.release()` method is not provided at all.
Usual design of Lock provides `.acquire`/`.release` pair and context manager.

But I hesitated to revise the behavior of current `.acquire` method because if it changes, users' program can be broken when users upgrade the aiozk package.
On the other hand, it would be understandable to change the API because aiozk is yet unstable version.

What do you think about changing the current behavior of `.acquire` method?
It would be nice to improve the design of Lock to be similar with asyncio.Lock.

**Comment by cybergrind** [comment]
@rhdxmr yeah, you're right about backward compatibility but the current implementation isn't what regular users expect from a modern library, so making `acquire` as a context manager was a terrible idea in a first place.
So, probably we're ready to raise a major version and release it as a beta, thus we will be free to break backward compatibility (like 1.0.0b1 and so on). 
This will be the first breaking change and probably we could do some refactoring and make codebase better in other places too.

**Comment by rhdxmr** [comment]
@cybergrind I rewrote this PR and pushed again.

By this modification, Lock provides async context manager, acquire/release pair and locked method.

Also SharedLock has .reader_lock and .writer_lock attributes. And also they supports async context manager, acquire/release pair and locked method.

This PR breaks current API of both Lock and SharedLock of which .acquire method returns async context manager.

Please review this PR.

File: aiozk/recipes/shared_lock.py [hunk 1]
```diff
@@ -1,27 +1,48 @@
...
-
 from .base_lock import BaseLock
 
 
 class SharedLock(BaseLock):
-    async def acquire_read(self, timeout=None):
-        deadline = Deadline(timeout)
-        result = None
-        while not result:
+    class LockProxy:
```
Comment (cybergrind) [inline aiozk/recipes/shared_lock.py:5]: Probably we can make this code cleaner if we allow passing `wait_with_party` parameter as the parameter of `Lock` and just use two locks.

File: aiozk/recipes/shared_lock.py [hunk 2]
```diff
@@ -1,27 +1,48 @@
...
+                self.is_locked = False
+                self.in_use = False
+
+        def locked(self):
+            return self.is_locked
+
+    def __init__(self, base_path):
+        super().__init__(base_path)
+
+        self.reader_lock = self.LockProxy(self, 'rlock', True)
```
Comment (cybergrind) [inline aiozk/recipes/shared_lock.py:47]: Probably it would be better if we use names from recipe https://zookeeper.apache.org/doc/r3.1.2/recipes.html#Shared+Locks, eg. `read` and `write`. Other libraries, that follows the same naming would be compatible in that case. I don't really expect it, but it won't create more confusion.
Also, could you please put the link to it into comment

File: aiozk/recipes/base_lock.py [hunk 3]
```diff
@@ -2,17 +2,30 @@
...
+
+    async def alert_session_lost(self):
+        await self._session_lost_future
+        log.warning(
+            'Session expired at some point, lock %s no longer acquired.', self)
+
+    async def _wait_in_line(self,
+                            znode_label,
+                            timeout=None,
+                            wait_with_party=False):
```
Comment (cybergrind) [inline aiozk/recipes/base_lock.py:24]: `blocked_by` was more explicit in two aspects:
1. users understand that we need to wait for another lock
2. users understand what exactly we're waiting for

In general, I don't think that `wait_in_line` really the best function that we should use. Probably we should have some function that allows to wait for lock without acquiring it. So read lock would be constructed like:

```python
read_lock = Lock('read', blocked_by=[Lock('write', only_wait=True)]
# or with | or &
read_lock = Lock('read') | Lock('write', only_wait=True)
```

File: aiozk/recipes/shared_lock.py (hunk 1 above)
Comment (rhdxmr) [reply aiozk/recipes/shared_lock.py:5]: OK. Reusing Lock class looks more simple.
But some extra keyword arguments should be added to it.
I will try it.

File: aiozk/recipes/shared_lock.py (hunk 2 above)
Comment (rhdxmr) [reply aiozk/recipes/shared_lock.py:47]: I will change the names.

File: aiozk/recipes/base_lock.py (hunk 3 above)
Comment (rhdxmr) [reply aiozk/recipes/base_lock.py:24]: OK. `blocked_by` will be returned.

However I do not understand the merit of the design you suggested.
IMO instantiating Lock objects and combining them may provide higher abstraction but I don't think it is needed now.
Maybe we can conduct refactoring later.
Also operator overloading does not catch my mind. It makes code be more complicated. And hard to get along with `grep`

Comment (cybergrind) [reply aiozk/recipes/base_lock.py:24]: In that case, I just want to make it more obvious that we're interacting with another `Lock`.

Operator overloading is probably too much, but ideally, it would be great if we can hide some low-level details with better abstractions. For example, lock code itself has only cryptic `wait_in_line` and `wait_in_line` has quite many repetitions of `while True: .. except SessionLost` pattern so it is hard to follow. I'm not ready to give a better design yet.

We definitely will merge this PR without major changes. But if you have any ideas about making code better - probably it is a good time to share them.

**Comment by rhdxmr** [comment]
@cybergrind

I applied your feedback and re-pushed the branch.

But I could not come up with better design of `BaseLock`. So still `.wait_in_line` is doing the same role for a lock mechanism.

Thanks,

**Comment by cybergrind** [comment]
@rhdxmr looks good, thank you!
